man/boomoney.1
.\" -*- coding: utf-8 -*-
.\" This file was generated automatically by tools/make_man.sh.
.TH BOOMONEY 1 "13 February 2020" "boomoney 2\&.0"
.SH NAME
boomoney \- import bank accounts into Microsoft Money
.SH SYNOPSIS
.B boomoney
[\-dqv] [\-b \fIbackends\fR] [\-cnfs] [\fIcommand\fR [\fIarguments\fR..]]
.br
.B boomoney
[\-\-help] [\-\-version]

.SH DESCRIPTION
.LP

Console application that imports bank accounts into Microsoft Money

.SS Supported websites:
* afer (Association française d'épargne et de retraite)
.br
* amazonstorecard (Amazon Store Card)
.br
* americanexpress (American Express)
.br
* amundi (Amundi)
.br
* apivie (Apivie)
.br
* aviva (Aviva)
.br
* axabanque (AXA Banque)
.br
* banqueaccord (Banque Accord)
.br
* banquepopulaire (Banque Populaire)
.br
* barclays (Barclays)
.br
* becm (Banque Européenne Crédit Mutuel)
.br
* bforbank (BforBank)
.br
* binck (Binck)
.br
* bnpcards (BNP Cartes Entreprises)
.br
* bnporc (BNP Paribas)
.br
* bnppere (BNP Épargne Salariale)
.br
* bolden (Bolden)
.br
* boursorama (Boursorama)
.br
* bp (La Banque Postale)
.br
* bred (Bred)
.br
* btpbanque (BTP Banque)
.br
* caels (Crédit Agricole \- Épargne Longue des Salariés)
.br
* caissedepargne (Caisse d'Épargne)
.br
* capeasi (AXA Épargne Salariale)
.br
* carrefourbanque (Carrefour Banque)
.br
* cic (CIC)
.br
* cices (CIC Épargne Salariale)
.br
* citibank (Citibank)
.br
* cmb (Crédit Mutuel de Bretagne)
.br
* cmes (Crédit Mutuel Épargne Salariale)
.br
* cmmc (Crédit Mutuel Massif Central)
.br
* cmso (Crédit Mutuel Sud\-Ouest)
.br
* cragr (Crédit Agricole)
.br
* creditcooperatif (Crédit Coopératif)
.br
* creditdunord (Crédit du Nord, Banque Courtois, Kolb, Nuger, Laydernier, Tarneaud, Société Marseillaise de Crédit)
.br
* creditdunordpee (Crédit du Nord Épargne Salariale)
.br
* creditmutuel (Crédit Mutuel)
.br
* degiro (De giro)
.br
* delubac (Banque Delubac & Cie)
.br
* erehsbc (HSBC Épargne Salariale)
.br
* esalia (Société Générale Épargne Salariale)
.br
* fortuneo (Fortuneo)
.br
* ganassurances (Gan Assurances)
.br
* gmf (GMF)
.br
* groupama (Groupama)
.br
* groupamaes (Groupama Épargne Salariale)
.br
* hsbc (HSBC France)
.br
* hsbchk (HSBC Hong Kong)
.br
* humanis (Humanis Épargne Salariale)
.br
* ing (ING France)
.br
* kiwibank (Kiwibank)
.br
* lcl (LCL)
.br
* lendosphere (Lendosphere)
.br
* linebourse (linebourse website)
.br
* lunchr (Lunchr)
.br
* myedenred (MyEdenRed)
.br
* n26 (Bank N26)
.br
* nalo (Nalo)
.br
* nef (La Nef)
.br
* netfinca (netfinca website)
.br
* oney (Oney)
.br
* paypal (PayPal)
.br
* pradoepargne (Prado Épargne Salariale)
.br
* s2e (Épargne Salariale)
.br
* societegenerale (Société Générale)
.br
* sogecartenet (Sogecarte Net)
.br
* spirica (Spirica)
.br
* suravenir (Assurance\-vie Suravenir à travers différents courtiers (assurancevie.com, linxea, ...))
.br
* themisbanque (Themis)
.br
* ticketscesu (Tickets CESU Edenred)
.br
* vicseccard (Victoria's Secret Angel Card)
.br
* wiseed (WiSEED)
.br
* yomoni (Yomoni)
.SH BOOMONEY COMMANDS
.TP
\fBadd_recipient\fR iban label
.br
Add a recipient to a backend.
.TP
\fBadvisor\fR
.br
Display advisors.
.TP
\fBbudgea\fR \fIUSERNAME\fR \fIPASSWORD\fR
.br
Export your bank accounts and transactions to Budgea.
.br

.br
Budgea is an online web and mobile application to manage your bank
.br
accounts. To avoid giving your credentials to this service, you can use
.br
this command.
.br

.br
https://www.budgea.com
.TP
\fBcoming\fR \fIID\fR [\fIEND_DATE\fR]
.br
Display future transactions.
.br

.br
If END_DATE is supplied, show all transactions until this date.
.br

.br
Default is limited to 10 results.
.TP
\fBdebug\fR
.br
Launch a debug Python shell
.TP
\fBhistory\fR \fIID\fR [\fIEND_DATE\fR]
.br
Display history of transactions.
.br

.br
If END_DATE is supplied, list all transactions until this date.
.br

.br
Default is limited to 10 results.
.TP
\fBinvestment\fR \fIID\fR
.br
Display investments of an account.
.TP
\fBlist\fR [\-\fIU\fR]
.br
List accounts.
.br
Use \-U to disable sorting of results.
.TP
\fBpocket\fR \fIID\fR
.br
Display pockets of an account.
.TP
\fBprofile\fR
.br
Display detailed information about person or company.
.TP
\fBrecipients\fR \fIACCOUNT\fR
.br
List recipients of ACCOUNT
.TP
\fBtransfer\fR [\fIACCOUNT\fR \fIRECIPIENT\fR \fIAMOUNT\fR [\fILABEL\fR [\fIEXEC_DATE\fR]]]
.br
Make a transfer beetwen two accounts
.br
\- ACCOUNT    the source account
.br
\- RECIPIENT  the recipient
.br
\- AMOUNT     amount to transfer
.br
\- LABEL      label of transfer
.br
\- EXEC_DATE  date when to execute the transfer
.SH WEBOOB COMMANDS
.TP
\fBbackends\fR [\fIACTION\fR] [\fIBACKEND_NAME\fR]...
.br
Select used backends.
.br

.br
ACTION is one of the following (default: list):
.br
* enable         enable given backends
.br
* disable        disable given backends
.br
* only           enable given backends and disable the others
.br
* list           list backends
.br
* add            add a backend
.br
* register       register a new account on a website
.br
* edit           edit a backend
.br
* remove         remove a backend
.br
* list\-modules   list modules
.TP
\fBcd\fR [\fIPATH\fR]
.br
Follow a path.
.br
".." is a special case and goes up one directory.
.br
"" is a special case and goes home.
.TP
\fBcondition\fR [\fIEXPRESSION\fR | off]
.br
If an argument is given, set the condition expression used to filter the results. See CONDITION section for more details and the expression.
.br
If the "off" value is given, conditional filtering is disabled.
.br

.br
If no argument is given, print the current condition expression.
.TP
\fBcount\fR [\fINUMBER\fR | off]
.br
If an argument is given, set the maximum number of results fetched.
.br
NUMBER must be at least 1.
.br
"off" value disables counting, and allows infinite searches.
.br

.br
If no argument is given, print the current count value.
.TP
\fBformatter\fR [list | \fIFORMATTER\fR [\fICOMMAND\fR] | option \fIOPTION_NAME\fR [on | off]]
.br
If a FORMATTER is given, set the formatter to use.
.br
You can add a COMMAND to apply the formatter change only to
.br
a given command.
.br

.br
If the argument is "list", print the available formatters.
.br

.br
If the argument is "option", set the formatter options.
.br
Valid options are: header, keys.
.br
If on/off value is given, set the value of the option.
.br
If not, print the current value for the option.
.br

.br
If no argument is given, print the current formatter.
.TP
\fBlogging\fR [\fILEVEL\fR]
.br
Set logging level.
.br

.br
Availables: debug, info, warning, error.
.br
* quiet is an alias for error
.br
* default is an alias for warning
.TP
\fBls\fR [\-d] [\-\fIU\fR] [\fIPATH\fR]
.br
List objects in current path.
.br
If an argument is given, list the specified path.
.br
Use \-U option to not sort results. It allows you to use a "fast path" to
.br
return results as soon as possible.
.br
Use \-d option to display information about a collection (and to not
.br
display the content of it). It has the same behavior than the well
.br
known UNIX "ls" command.
.br

.br
Default is limited to 40 results.
.TP
\fBquit\fR
.br
Quit the application.
.TP
\fBselect\fR [\fIFIELD_NAME\fR]... | "$direct" | "$full"
.br
If an argument is given, set the selected fields.
.br
$direct selects all fields loaded in one http request.
.br
$full selects all fields using as much http requests as necessary.
.br

.br
If no argument is given, print the currently selected fields.

.SH OPTIONS
.TP
\fB\-\-version\fR
show program's version number and exit
.TP
\fB\-h\fR, \fB\-\-help\fR
show this help message and exit
.TP
\fB\-b BACKENDS\fR, \fB\-\-backends=BACKENDS\fR
what backend(s) to enable (comma separated)
.TP
\fB\-e EXCLUDE_BACKENDS\fR, \fB\-\-exclude\-backends=EXCLUDE_BACKENDS\fR
what backend(s) to exclude (comma separated)
.TP
\fB\-I\fR, \fB\-\-insecure\fR
do not validate SSL
.TP
\fB\-\-nss\fR
Use NSS instead of OpenSSL
.TP
\fB\-\-auto\-update\fR
Automatically check for updates when a bug in a module is encountered

.SH LOGGING OPTIONS
.TP
\fB\-d\fR, \fB\-\-debug\fR
display debug messages. Set up it twice to more verbosity
.TP
\fB\-q\fR, \fB\-\-quiet\fR
display only error messages
.TP
\fB\-v\fR, \fB\-\-verbose\fR
display info messages
.TP
\fB\-\-logging\-file=LOGGING_FILE\fR
file to save logs
.TP
\fB\-a\fR, \fB\-\-save\-responses\fR
save every response
.TP
\fB\-\-export\-session\fR
log browser session cookies after login

.SH RESULTS OPTIONS
.TP
\fB\-c CONDITION\fR, \fB\-\-condition=CONDITION\fR
filter result items to display given a boolean expression. See CONDITION section
for the syntax
.TP
\fB\-n COUNT\fR, \fB\-\-count=COUNT\fR
limit number of results (from each backends)
.TP
\fB\-s SELECT\fR, \fB\-\-select=SELECT\fR
select result item keys to display (comma separated)

.SH FORMATTING OPTIONS
.TP
\fB\-f FORMATTER\fR, \fB\-\-formatter=FORMATTER\fR
select output formatter (csv, htmltable, json, json_line, list, multiline,
simple, table, webkit)
.TP
\fB\-\-no\-header\fR
do not display header
.TP
\fB\-\-no\-keys\fR
do not display item keys
.TP
\fB\-O OUTFILE\fR, \fB\-\-outfile=OUTFILE\fR
file to export result

.SH BOOMONEY OPTIONS
.TP
\fB\-F\fR, \fB\-\-force\fR
forces the retrieval of transactions (10 maximum), otherwise retrieves only the
transactions newer than the previous retrieval date
.TP
\fB\-A ACCOUNT\fR, \fB\-\-account=ACCOUNT\fR
retrieves only the specified account. By default, all accounts are retrieved
.TP
\fB\-N\fR, \fB\-\-noimport\fR
no import. Generates the files, but they are not imported in MSMoney. Last
import dates are not modified
.TP

ChangeLog
Weboob 2.0 (2020-02-13)

	Applications: boobank
	* keep Value for TransferStep and AddRecipientStep
	* declare utf-8 in OFX formatter
	* give account object to iter_transfer_recipient

	Applications: boobill
	* using a persistent storage for browser state

	Applications: boobsize
	* fix name

	Applications: translaboob
	* use babel to detect language names and pass ISO-639 codes

	Applications: weboobdebug
	* weboob-debug: fix name

	Applications: weboobmain
	* move scripts/weboob to a dedicated dir in weboob/applications

	Applications: weboobrepos
	* weboob-repos: port to python3

	Browser
	* set a parent logger to AbstractElement
	* Allows SeleniumBrowser to use MANUAL proxies
	* AUTHENTICATION_METHODS set in init
	* delete prefix 'handle_' on authentication methods
	* clean authentication config keys after login
	* make variable name more understandable
	* make get_expire stronger
	* TWOFA_DURATION for the state expire date
	* better names
	* set expire from logged_date
	* make authentication methods more flexible
	* handle both 2FA and regular login methods
	* add possibility to clear specific cookies
	* handle NeedInteractiveFor2FA
	* add documentation about otp, sms and resume
	* add TwoFactorBrowser to handle multiple 2FA
	* Set default logging level to ERROR for selenium
	* Added the ability to overrride *Elements' env copy
	* Allow passing custom headers in Form.submit()
	* set attribute 'weboob' on browser
	* Normalize encodings
	* BrowserTooManyRequests: Remove class attributes for date helpers
	* BrowserTooManyRequests: Raise a TypeError if the next_try is not a datetime
	* BrowserTooManyRequests: Check whether the next_try is None after the date check
	* BrowserTooManyRequests: Add message parameter
	* [weboob/browser/browser+pages] Fixed doctest BASEURL
	* add some helpers to TooManyRequests
	* add rate limiting exception
	* add build_refresh_token_parameters method
	* python3 compatibility
	* send redirect_uri when using the refresh_token
	* allow multiple inheritance with AbstractBrowser
	* Implemented wait_xpath_invisible
	* Avoid crash when no BASEURL is set
	* add klass attribute to decorated @method
	* give state to the module at initialization
	* introduce BrowserParamURL for passing params implictly
	* Implement the empty_xpath attribute
	* accept new parameter proxy_headers (but don't use it)
	* Browser takes a new parameter 'proxy_headers'
	* if use of refresh_token fails, forget it
	* support when token_type key is missing
	* if access_token_expire is None, the access_token has no expiration
	* in case we got a 401 response, forget the access_token
	* fix compatibility with python3
	* Handle generic error for bad url callback
	* Repair refresh token
	* StatesMixin cookies should be str
	* OAuth2PKCEMixin codes shouldn't be bytes
	* Rewrite/improve PagesBrowser example

	Browser: Filters
	* Add the format .Italian() to CleanDecimal
	* New keyword argument to remove accented characters on CleanText filter.
	* Changed Capitalize filter to Title
	* Updated & Reordered __all__
	* Add dates to FormValue
	* Add color to FormValue
	* Add number/range type to FormValue, and tests
	* MultiJoin accepts types other than str
	* Regexp can use regex library
	* Allow all dateutil.parser arguments in DateTime
	* Strict dates by default
	* Catch incomplete dates
	* Explicit error message
	* Add MultiJoin
	* Join can use default argument
	* Add doctests to Join
	* TableCell(colspan=True) is now the only behavior desired
	* add FromTimestamp to convert to datetime
	* Fix error message for FormValue
	* Add new types for FormValue
	* Add MapIn filter, that allow the pattern of a selected value to another value using a dict.
	* Reorganize imports / add missing to __all__
	* Add Coalesce

	Capabilities
	* rename accepted_doc_types attribute of CapDocument
	* Add KIID to DocumentTypes
	* Added new capability CapBands
	* Keep only PER
	* add PER account type
	* HousingPhoto inherits BaseImage
	* use PostalAddress and GeoCoords types where possible
	* introduce weboob.capabilities.address for a few object types
	* Add firstname and lastname field
	* add capability_to_string function to get capability name
	* Add new SAR currency
	* fix missing TransferInvalidOTP export
	* add TransferInvalidOTP exception
	* use StringField standard default value for transfer beneficiary label and number
	* Indentation cosmetic changes
	* Add new Investment attributes
	* Add new document type 'income_tax'
	* Add field 'original_gross_amount' to Transaction
	* Add field 'gross_amount' to Transaction
	* Remove performance_history default value
	* Added new attributes for employee savings accounts
	* Add RecaptchaV3Job to captcha capability
	* add support for commission in currency in bank/Transaction

weboob/applications/boomoney/boomoney.py
# -*- coding: utf-8 -*-

# Copyright(C) 2009-2011  Romain Bignon, Christophe Benz
#
# This file is part of weboob.
#
# weboob is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# weboob is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with weboob. If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

# start with:
# set PYTHONPATH=D:\Dropbox\Projets\boomoney
# python d:\Dropbox\Projets\boomoney\scripts\boomoney -N


from threading import Thread, Lock
import copy
import sys
from io import StringIO
import os
import re
import subprocess
import datetime
from optparse import OptionGroup

import shutil
from colorama import init, Fore, Style

from weboob.tools.compat import unicode
from weboob.exceptions import BrowserUnavailable
from weboob.capabilities.bank import AccountNotFound, AccountType
from weboob.applications.boobank import Boobank
from weboob.applications.boobank.boobank import OfxFormatter
from weboob.tools.application.formatters.simple import SimpleFormatter


__all__ = ['Boomoney']

printMutex = Lock()
numMutex = Lock()
backupMutex = Lock()


class MoneyOfxFormatter(OfxFormatter):
    def start_format(self, **kwargs):
        self.seen = set()
        # MSMoney only supports CHECKING accounts
        t = kwargs['account'].type
        kwargs['account'].type = AccountType.CHECKING
        super(MoneyOfxFormatter, self).start_format(**kwargs)
        kwargs['account'].type = t

    def format_obj(self, obj, alias):
        cat = obj.category
        obj.category = obj.raw
        result = super(MoneyOfxFormatter, self).format_obj(obj, alias)
        obj.category = cat
        return result

    def output(self, formatted):
        if self.outfile != sys.stdout:
            self.outfile.write(formatted + os.linesep)
        else:
            super(MoneyOfxFormatter, self).output(formatted)


class ListFormatter(SimpleFormatter):
    def output(self, formatted):
        if self.outfile != sys.stdout:
            self.outfile.write(formatted + os.linesep)
        else:
            super(ListFormatter, self).output(formatted)


class BoobankNoBackend(Boobank):
    EXTRA_FORMATTERS = {'ops_list': MoneyOfxFormatter}
    COMMANDS_FORMATTERS = {'history': 'ops_list'}

    def load_default_backends(self):
        pass

    def bcall_error_handler(self, backend, error, backtrace):
        handled = False
        if isinstance(error, BrowserUnavailable):
            handled = True
            self.error = True
        if isinstance(error, AccountNotFound):
            handled = True
            self.error = True
        if isinstance(error, NotImplementedError):
            handled = True
            self.error = False
        if not handled:
            self.error = True
            self.weboob.logger.error("Unsupported error %s in BoobankNoBackend" % type(error))
        return super(Boobank, self).bcall_error_handler(backend, error, backtrace)


class HistoryThread(Thread):
    def __init__(self, boomoney, account):
        Thread.__init__(self)
        self.boomoney = boomoney
        self.account = account
        self.disabled = boomoney.config.get(account, 'disabled', default=False)
        self.date_min = boomoney.config.get(account, 'date_min', default='')
        self.last_date = boomoney.config.get(account, 'last_date', default='')

    @property
    def label(self):
        if self.account in self.boomoney.labels:
            return self.boomoney.labels[self.account]
        else:
            return self.boomoney.config.get(self.account, 'label', default='')

    def dumpTransaction(self, output, fields, field):
        output.write("<STMTTRN>\n")
        if "DTUSER" in field and "DTPOSTED" in field and not field["DTUSER"] == field["DTPOSTED"]:
            # the payment date is a deferred payment
            # MSMoney takes DTPOSTED, which is the payment date
            # I prefer to have the date of the operation, so I set DTPOSTED
            # as DTUSER
            field["DTPOSTED"] = field["DTUSER"]
        for f in fields.strip().split(" "):
            value = field[f]
            if f == "NAME":
                if value == "":
                    # MSMoney does not support empty NAME field
                    value = "</NAME>"
                else:
                    # MSMoney does not support NAME field longer than 64
                    value = value[:64]
            output.write("<%s>%s\n" % (f, value))
        output.write("</STMTTRN>\n")

    def run(self):

        now = datetime.datetime.now().strftime("%Y-%m-%d")

        if self.boomoney.options.force:
            from_date = self.date_min
        else:
            from_date = self.last_date

        if from_date >= now:
            self.boomoney.print(Style.BRIGHT + "%s (%s): Last import date is %s, no need to import again..." % (
                self.account, self.label, self.last_date) + Style.RESET_ALL)
            return

        boobank = self.boomoney.createBoobank(self.account)
        if boobank is None:
            with numMutex:
                self.boomoney.importIndex = self.boomoney.importIndex + 1
            return

        boobank.stderr = StringIO()
        boobank.stdout = boobank.stderr
        id, backend = self.account.split("@")
        module_name, foo = boobank.weboob.backends_config.get_backend(backend)
        moduleHandler = "%s.bat" % os.path.join(os.path.dirname(self.boomoney.getMoneyFile()), module_name)
        self.boomoney.logger.info("Starting history of %s (%s)..." % (self.account, self.label))

        MAX_RETRIES = 3
        count = 0
        found = False
        content = ''
        boobank.error = False
        while count <= MAX_RETRIES and not (found and not boobank.error):
            boobank.options.outfile = StringIO()
            boobank.error = False

            # executing history command
            boobank.onecmd("history " + self.account + " " + from_date)

            if count > 0:
                self.boomoney.logger.info("Retrying %s (%s)... %i/%i" % (self.account, self.label, count, MAX_RETRIES))
            found = re.match(r'^OFXHEADER:100', boobank.options.outfile.getvalue())
            if found and not boobank.error:
                content = boobank.options.outfile.getvalue()
            boobank.options.outfile.close()
            count = count + 1
        if content == '':
            # error occurred
            with numMutex:
                self.boomoney.importIndex = self.boomoney.importIndex + 1
                index = self.boomoney.importIndex
            self.boomoney.logger.error("(%i/%i) %s (%s): %saborting after %i retries.%s" % (
                index, len(self.boomoney.threads),
                self.account,
                self.label,
                Fore.RED + Style.BRIGHT,
                MAX_RETRIES,
                Style.RESET_ALL))
            return

        # postprocessing of the ofx content to match MSMoney expectations
        content = re.sub(r'<BALAMT>Not loaded', r'<BALAMT></BALAMT>', content)
        input = StringIO(content)
        output = StringIO()
        field = {}
        fields = ' '
        for line in input:
            if re.match(r'^OFXHEADER:100', line):
                inTransaction = False
            if re.match(r'^<STMTTRN>', line):
                inTransaction = True
            if not inTransaction:
                output.write(line)
            if re.match(r'^</STMTTRN>', line):
                # MSMoney expects CHECKNUM instead of NAME for CHECK transactions
                if "TRNTYPE" in field and field["TRNTYPE"] == "CHECK":
                    if "NAME" in field and unicode(field["NAME"]).isnumeric():
                        field["CHECKNUM"] = field["NAME"]
                        del field["NAME"]
                        fields = fields.replace(' NAME ', ' CHECKNUM ')

                # go through specific backend process if any
                IGNORE = False
                NEW = None
                origfields = fields
                origfield = field.copy()
                if os.path.exists(moduleHandler):
                    self.boomoney.logger.info("Calling backend handler %s..." % moduleHandler)
                    # apply the transformations, in the form
                    # field_NAME=...
                    # field_MEMO=...
                    # field=...

                    cmd = 'cmd /C '
                    for f in field:
                        value = field[f]
                        cmd = cmd + 'set field_%s=%s& ' % (f, value)

                    cmd = cmd + '"' + moduleHandler + '"'
                    result = subprocess.check_output(cmd.encode(sys.stdout.encoding))

                    for line in re.split(r'[\r\n]+', result):
                        if not line == "":
                            f, value = line.split("=", 1)

                            if f == "IGNORE":
                                IGNORE = True
                            elif f == "NEW":
                                NEW = value
                            elif f.startswith('field_'):
                                f = re.sub(r'^field_', '', f)
                                if value == "":
                                    if f in field:
                                        del field[f]
                                    fields = re.sub(" " + f + " ", " ", fields)
                                else:
                                    field[f] = value
                                    if f not in fields.strip().split(" "):
                                        # MSMoney does not like when CHECKNUM is after MEMO
                                        if f == "CHECKNUM":
                                            fields = fields.replace("MEMO", "CHECKNUM MEMO")
                                        else:
                                            fields = fields + f + " "

                if not IGNORE:
                    # dump transaction
                    self.dumpTransaction(output, fields, field)

                    if NEW is not None:
                        for n in NEW.strip().split(" "):
                            fields = origfields
                            field = origfield.copy()
                            field["FITID"] = origfield["FITID"] + "_" + n
                            for line in re.split(r'[\r\n]+', result):
                                if not line == "":
                                    f, value = line.split("=", 1)

                                    if f.startswith(n + '_field_'):
                                        f = re.sub(r'^.*_field_', '', f)
                                        field[f] = value
                                        if f not in fields.strip().split(" "):
                                            fields = fields + f + " "
                            # dump secondary transaction
                            self.dumpTransaction(output, fields, field)

                inTransaction = False
            if inTransaction:
                if re.match(r'^<STMTTRN>', line):
                    field = {}
                    fields = ' '
                else:
                    t = line.split(">", 1)
                    v = re.split(r'[\r\n]', t[1])
                    field[t[0][1:]] = v[0]
                    fields = fields + t[0][1:] + ' '

        ofxcontent = output.getvalue()
        stderrcontent = boobank.stderr.getvalue()
        input.close()
        output.close()
        boobank.stderr.close()

        if self.boomoney.options.display:
            self.boomoney.print(Style.BRIGHT + ofxcontent + Style.RESET_ALL)

        nbTransactions = ofxcontent.count('<STMTTRN>')

        # create ofx file
        fname = re.sub(r'[^\w@\. ]', '_', self.account + " " + self.label)
        ofxfile = os.path.join(self.boomoney.getDownloadsPath(), fname + ".ofx")
        with open(ofxfile, "w") as ofx_file:
            ofx_file.write(re.sub(r'\r\n', r'\n', ofxcontent.encode(sys.stdout.encoding)))

        with numMutex:
            self.boomoney.write(stderrcontent)
            self.boomoney.importIndex = self.boomoney.importIndex + 1
            index = self.boomoney.importIndex
        if not (self.boomoney.options.noimport or nbTransactions == 0):
            self.boomoney.backupIfNeeded()
        with printMutex:
            if self.boomoney.options.noimport or nbTransactions == 0:
                if nbTransactions == 0:
                    print(Style.BRIGHT + '(%i/%i) %s (%s) (no transaction).' % (
                        index, len(self.boomoney.threads),
                        self.account,
                        self.label
                    ) + Style.RESET_ALL)
                else:
                    print(Fore.GREEN + Style.BRIGHT + '(%i/%i) %s (%s) (%i transaction(s)).' % (
                        index, len(self.boomoney.threads),
                        self.account,
                        self.label,
                        nbTransactions
                    ) + Style.RESET_ALL)
            else:
                # import into money
                print(Fore.GREEN + Style.BRIGHT + '(%i/%i) Importing "%s" into MSMoney (%i transaction(s))...' % (
                    index, len(self.boomoney.threads),
                    ofxfile,
                    nbTransactions
                ) + Style.RESET_ALL)
        if not self.boomoney.options.noimport:
            if nbTransactions > 0:
                subprocess.check_call('"%s" %s' % (
                    os.path.join(self.boomoney.getMoneyPath(), "mnyimprt.exe"),
                    ofxfile))
            self.last_date = now


class Boomoney(Boobank):
    APPNAME = 'boomoney'
    VERSION = '2.1'
    COPYRIGHT = 'Copyright(C) 2018-YEAR Bruno Chabrier'
    DESCRIPTION = "Console application that imports bank accounts into Microsoft Money"
    SHORT_DESCRIPTION = "import bank accounts into Microsoft Money"

    EXTRA_FORMATTERS = {'list': ListFormatter}
    COMMANDS_FORMATTERS = {'list': 'list'}

    def __init__(self):
        super(Boobank, self).__init__()
        self.importIndex = 0
        application_options = OptionGroup(self._parser, 'Boomoney Options')
        application_options.add_option('-F', '--force', action='store_true', help='forces the retrieval of transactions (10 maximum), otherwise retrieves only the transactions newer than the previous retrieval date')
        application_options.add_option('-A', '--account', help='retrieves only the specified account. By default, all accounts are retrieved')
        application_options.add_option('-N', '--noimport', action='store_true', help='no import. Generates the files, but they are not imported in MSMoney. Last import dates are not modified')
        application_options.add_option('-D', '--display', action='store_true', help='displays the generated OFX file')
        application_options.add_option('-P', '--parallel', action='store_true', help='retrieves all accounts in parallel instead of one by one (experimental)')
        self._parser.add_option_group(application_options)
        self.labels = dict()

    def print(self, *args):
        with printMutex:
            print(*args)

    def write(self, *args):
        with printMutex:
            sys.stdout.write(*args)

    def createBoobank(self, account):
        accountId, backendName = account.split("@")

        if not self.weboob.backends_config.backend_exists(backendName):
            self.logger.warning("Unknown backend '%s' of account '%s' (not found in backends)" % (backendName, account))
            return None

        # create a Boobank instance
        boobank = BoobankNoBackend()
        boobank.options = copy.copy(self.options)

        moduleName = self.weboob.backends_config._read_config().get(backendName, "_module")
        module = self.weboob.modules_loader.loaded[moduleName]
        backend = self.weboob.backend_instances[backendName]

        params = {}
        for param in backend.config:
            params[param] = backend.config[param].get()
        dedicatedBackendInstanceName = "backend instance for " + account
        boobank.APP_NAME = "boobank app for " + account
        instance = module.create_instance(self.weboob, dedicatedBackendInstanceName, params, storage=boobank.create_storage())

        boobank.enabled_backends = set()
        boobank.enabled_backends.add(instance)
        boobank.weboob.backend_instances[dedicatedBackendInstanceName] = instance

        boobank.selected_fields = ["$full"]
        boobank.formatter = self.formatter

        boobank._interactive = False
        return boobank

    def getHistory(self, account):
        t = HistoryThread(self, account)
        return t

    def getDownloadsPath(self):
        if not hasattr(self, '_downloadsPath'):
            s = subprocess.check_output(
                'reg query "HKEY_CURRENT_USER\\Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\User Shell Folders" /v "{374DE290-123F-4565-9164-39C4925E467B}"')
            t = re.sub(r'^(.|\r|\n)+REG_EXPAND_SZ\s+([^\n\r]+)(.|\r|\n)*$', r'\2', s)
            self._downloadsPath = os.path.expandvars(t).decode('CP850')
        return self._downloadsPath

    def getMoneyPath(self):
        if not hasattr(self, '_moneyPath'):
            s = subprocess.check_output('reg query HKEY_CLASSES_ROOT\\money\\Shell\\Open\\Command /ve')
            t = re.sub(r'^(.|\r|\n)+REG_SZ\s+([^\n\r]+)(.|\r|\n)*$', r'\2', s)
            self._moneyPath = os.path.expandvars(os.path.dirname(t)).decode('CP850')
        return self._moneyPath

    def getMoneyFile(self):
        if not hasattr(self, '_moneyFile'):
            s = subprocess.check_output('reg query HKEY_CURRENT_USER\\Software\\Microsoft\\Money\\14.0 /v CurrentFile')
            t = re.sub(r'^(.|\r|\n)+REG_SZ\s+([^\n\r]+)(.|\r|\n)*$', r'\2', s)
            self._moneyFile = os.path.expandvars(t).decode('CP850')
        return self._moneyFile

    def backupIfNeeded(self):
        if not (hasattr(self, '_backupDone') and self._backupDone):
            with backupMutex:
                # redo the test in mutual exclusion
                if not (hasattr(self, '_backupDone') and self._backupDone):
                    file = self.getMoneyFile()
                    filename = os.path.splitext(os.path.basename(file))[0]
                    dir = os.path.dirname(file)
                    self.print(Fore.YELLOW + Style.BRIGHT + "Creating backup of %s..." % file + Style.RESET_ALL)
                    target = os.path.join(dir, filename + datetime.datetime.now().strftime("_%Y_%m_%d_%H%M%S.mny"))
                    shutil.copy2(file, target)
                    self._backupDone = True

    def save_config(self):
        for t in self.threads:
            self.config.set(t.account, 'label', t.label)
            self.config.set(t.account, 'disabled', t.disabled)
            self.config.set(t.account, 'date_min', t.date_min)
            self.config.set(t.account, 'last_date', t.last_date)

        self.config.save()

    def getList(self):
        self.onecmd("select id label")
        self.options.outfile = StringIO()
        self.onecmd("list")
        listContent = self.options.outfile.getvalue()
        self.options.outfile.close()
        self.print(Style.BRIGHT + "Accounts:%s----------%s%s----------" % (
            os.linesep,
            os.linesep,
            listContent) + Style.RESET_ALL)
        for line in listContent.split(os.linesep):
            if not line == "":
                idspec, labelspec = line.split("\t")
                notusedid, id = idspec.split("=")
                notusedlabel, label = labelspec.split("=")
                self.labels[id] = label

    def checkNew(self):
        new = set()
        for account in self.labels:
            if account not in self.config.config.sections():
                new.add(HistoryThread(self, account))
        return new

    def main(self, argv):

        init()

        self.load_config()

        self._interactive = False

        self.threads = set()

        self.logger.info(self.config.config.sections())
        for account in self.config.config.sections():
            if self.options.account == None or account == self.options.account:
                if self.config.config.getboolean(account, "disabled") == False:
                    # time.sleep(3)
                    self.threads.add(self.getHistory(account))

        if self.options.parallel:
            self.print(Fore.MAGENTA + Style.BRIGHT + "Starting %i history threads..." % len(self.threads) + Style.RESET_ALL)
            for t in self.threads:
                t.start()
            self.getList()
            for t in self.checkNew():
                t.start()
                self.threads.add(t)
            self.print(Fore.MAGENTA + Style.BRIGHT + "Waiting for %i threads to complete..." % len(self.threads) + Style.RESET_ALL)
            for t in self.threads:
                t.join()
        else:
            self.getList()
            for t in self.checkNew():
                self.threads.add(t)
            for t in self.threads:
                t.start()
                t.join()

        self.save_config()
        return
